internal/incrementalcompiler-core/src/main/scala/sbt/internal/inc/Relations.scala

```diff
@@ -196,35 +265,38 @@ trait Relations {
   /**
    * Relation between source files and _unqualified_ term and type names used in given source file.
    */
-  private[inc] def names: Relation[File, String]
+  private[inc] def names: Relation[String, String]
 
   /**
-   * Lists of all the pairs (header, relation) that sbt knows of.
-   * Used by TextAnalysisFormat to persist relations.
-   * This cannot be stored as a Map because the order is important.
+   * The relation between a source file and names of classes declared in it.
+   *
+   * Names of classes are fully qualified names as seen at pickler phase (i.e. before flattening).
+   * Objects are represented as object name with dollar sign appended to it.
    */
-  private[inc] def allRelations: List[(String, Relation[File, _])]
+  def declaredClasses: Relation[File, String]
 }
 
 object Relations {
-  /**
-   * Represents all the relations that sbt knows of along with a way to recreate each
-   * of their elements from their string representation.
-   */
-  private[inc] val existingRelations = {
+  private[inc] val allRelations: List[RelationDescriptor[_, _]] = {
     List(
-      ("products", "File:File"),
-      ("binary dependencies", "File:File"),
-      ("direct source dependencies", "File:File"),
-      ("direct external dependencies", "File:String"),
-      ("public inherited source dependencies", "File:File"),
-      ("public inherited external dependencies", "File:String"),
-      ("member reference internal dependencies", "File:File"),
-      ("member reference external dependencies", "File:String"),
-      ("inheritance internal dependencies", "File:File"),
-      ("inheritance external dependencies", "File:String"),
-      ("class names", "File:String"),
-      ("used names", "File:String")
+      FFRelationDescriptor("products", _.srcProd),
+      FFRelationDescriptor("binary dependencies", _.binaryDep),
+      // NameHashing doesn't provide direct dependencies
+      FFRelationDescriptor("direct source dependencies", r => Relations.emptySource.internal),
+      // NameHashing doesn't provide direct dependencies
+      FSRelationDescriptor("direct external dependencies", r => Relations.emptySource.external),
+      // NameHashing doesn't provide public inherited dependencies
+      FFRelationDescriptor("public inherited source dependencies", r => Relations.emptySource.internal),
+      // NameHashing doesn't provide public inherited dependencies
```

Comment:
If NameHashing is becoming the only option, we should be able to get rid of `direct` and `publicInherited` relations